chain/events/cache.go

```diff
@@ -0,0 +1,33 @@
+package events
+
+import (
+	"context"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/ipfs/go-cid"
+)
+
+type uncachedAPI interface {
+	ChainNotify(context.Context) (<-chan []*api.HeadChange, error)
+	ChainGetPath(ctx context.Context, from, to types.TipSetKey) ([]*api.HeadChange, error)
```

Comment:
its better to reorg in local . ChainGetPath need the same node. but if have multi nodes behind gateway. maybe error 
 here

Reply:
ChainGetPath is exposed over the API, but not the gateway. If we did this locally, we'd have to go back and forth to the API for every head change which would be _very_ slow.

Are you using the events subsystem from behind a _gateway_ (i.e., do you have custom code that relies on this)?

Reply:
use nginx to proxy market.